jenkins/agent-ctl.sh

```
@@ -0,0 +1,305 @@
+  KIND='heavy'
+elif [[ "${INSTANCE}" =~ pr- ]]; then
+  KIND='pr'
+else
+  KIND=
+fi
+
+case "${KIND}" in
+  light)
+    # Current experiment:
+    # 20 executors, n1-highmem-16, 500G pd-standard
+    # Results:
+    # 5.31 cores, 70G ram, <500 write IOPs, <50MB/s write
+    # Alernate experiment:
+    # 2 executors, n1-highmem-2, 100G pd-standard
+    # Results:
+    # 0.2 cores, 1.3G ram, low IOPs (200 IOP spikes), low write (30MB/s spikes)
+    DISK_SIZE='500GB'
+    DISK_TYPE='pd-standard'
+    MACHINE_TYPE='n1-highmem-16'
+    ;;
+  heavy)
+    # TODO(fejta): optimize these values
+    DISK_SIZE='200GB'
+    DISK_TYPE='pd-standard'
+    MACHINE_TYPE='n1-standard-16'
+    ;;
+  pr)
+    # TODO(fejta): optimize these values
+    DISK_SIZE='500GB'
+    DISK_TYPE='pd-ssd'
+    MACHINE_TYPE='n1-highmem-32'
+    ;;
+  *)
+    ;;
+esac
+
+if [[ -z "${REAL}" ]]; then
+  DISK_SIZE='200GB'
+  DISK_TYPE='pd-standard'
+  MACHINE_TYPE='n1-standard-1'
+  read -p "Using ${MACHINE_TYPE} for testing. Continue [Y/n]: " ans
+  if [[ ! "${ans}" =~ '^[yY]' ]]; then
+    echo 'Add --real'
+    exit 1
+  fi
+fi
+
+auto-agent() {
+  echo "Automatically creating ${INSTANCE}..."
+  detatch-agent
+  delete-agent
+  create-agent
+  update-agent
+  reboot-agent
+  attach-agent
+}
+
+tunnel-to-master() {
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      sleep 1
+    else
+      break
+    fi
+  done
+  sudo netstat -anp | grep 8080 && echo "8080 already used" && exit 1 \
+    || echo "Tunneling to ${MASTER}..."
+  gcloud compute ssh "${MASTER}" --ssh-flag='-L8080:localhost:8080' sleep 5 &
+  for i in {1..10}; do
+    if sudo netstat -anp | grep 8080 > /dev/null 2>&1 ; then
+      break
+    fi
+    sleep 1
+  done
+  sudo netstat -anp | grep 8080 > /dev/null 2>&1 || exit 1
+}
+
+
+master-change() {
+  tunnel-to-master
+  cmd="${1}"
+  ini="${HOME}/jenkins-master-creds.ini"  # /user/<user>/configure && show api token
```

should make sure this file exists and fail early if it doesn't